### PR TITLE
[RISCV] Add a MIR pass to reassociate shXadd, add, and slli to form more shXadd.

### PR DESCRIPTION
**Pull request by topperc**
This reassociates patterns like (sh3add Z, (add X, (slli Y, 6)))
into (sh3add (sh3add Y, Z), X).
    
This improves a pattern that occurs in 531.deepsjeng_r. Reducing
the dynamic instruction count by 0.5%.
    
This may be possible to improve in SelectionDAG, but given the special
cases around shXadd formation, it's not obvious it can be done in a
robust way without adding multiple special cases.
    
I've used a GEP with 2 indices because that mostly closely resembles
the motivating case. Most of the test cases are the simplest GEP case.
One test has a logical right shift on an index which is closer to
the deepsjeng code. This requires special handling in isel to reverse
a DAGCombiner canonicalization that turns a pair of shifts into
(srl (and X, C1), C2).
    
See also #85734 which had a hacky version of a similar optimization.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVOptZba.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      if (!Add || Add->getOpcode() != RISCV::ADD || Add->getParent() != &MBB ||
+          !MRI.hasOneNonDBGUse(UnshiftedReg))
+        continue;
+
+      Register AddReg0 = Add->getOperand(1).getReg();
+      Register AddReg1 = Add->getOperand(2).getReg();
+
+      MachineInstr *InnerShift;
+      Register X;
+      if (findShift(AddReg0, MBB, InnerShift, MRI))
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVOptZba.cpp:114]: I guess some optimization opportunities will be missed when both operands are shifts.
An example: `(sh3add Z, (add (slli X, 7), (slli Y, 6)))`

File: llvm/lib/Target/RISCV/RISCVOptZba.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      else if (findShift(AddReg1, MBB, InnerShift, MRI))
+        X = AddReg0;
+      else
+        continue;
+
+      unsigned InnerShiftAmt = InnerShift->getOperand(2).getImm();
+
+      // The inner shift amount must be at least as large as the outer shift
+      // amount.
+      if (OuterShiftAmt > InnerShiftAmt)
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVOptZba.cpp:125]: I think we can still reassociate this, but that can be a separate patch.  :)

File: llvm/lib/Target/RISCV/RISCVOptZba.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+//
+//===---------------------------------------------------------------------===//
+//
+// This pass attempts to reassociate expressions like
+//   (sh3add Z, (add X, (slli Y, 6)))
+// To
+//   (sh3add (sh3add Y, Z), X)
+//
+// This reduces number of instructions needing by spreading the shift amount
+// across to 2 sh3adds. This can be generalized to sh1add/sh2add and other
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVOptZba.cpp:15]: The second part of this comment looks stale?

**Review by preames — COMMENTED** [review]
A bit of context for anyone reading along.

Craig and I spent a bunch of time chatting about related issues last week.  I'm going to continue with a set of patches to try and both generate shNadds more often, and (possibly) restructure how these are handled in DAG combine.

I suggested that Craig land a variant of the original patch (i.e. this one), and that we use that as our baseline for further iteration.  We may end up reversing course here fairly quickly, but I don't see any significant downside to having landed this if that turns out be the case.

File: llvm/lib/Target/RISCV/RISCVOptZba.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+};
+
+} // end anonymous namespace
+
+char RISCVOptZba::ID = 0;
+INITIALIZE_PASS(RISCVOptZba, DEBUG_TYPE, RISCV_OPT_ZBA_NAME, false, false)
+
+FunctionPass *llvm::createRISCVOptZbaPass() { return new RISCVOptZba(); }
+
+static bool findShift(Register Reg, const MachineBasicBlock &MBB,
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVOptZba.cpp:56]: Yes. I originally had it returning a bunch of output parameters instead of the Shift instruction. Then I realized I needed the shift instruction in order to erase it.

File: llvm/lib/Target/RISCV/RISCVOptZba.cpp (hunk 3 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVOptZba.cpp:15]: I meant it can be generalized relative to the specific example in the comment, but I admit that was probably poorly worded.

File: llvm/lib/Target/RISCV/RISCVOptZba.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVOptZba.cpp:125]: That would have to be a different reassociation of some kind. Do you have an example?

**Comment by topperc** [comment]
I believe I have this working in the MachineCombiner now. I'll post that version shortly.
